File: tools/sdk/sdk/python/src/vibecode_agents/agents/code_review.py
"""
Code Review Agent

Automated code review agent with security, performance,
and quality analysis capabilities.
"""

import logging
from pathlib import Path
from typing import Dict, List, Optional

from vibecode_agents.client import AgentClient
from vibecode_agents.models import AgentType, ModelType, StartAgentRequest
from vibecode_agents.tools import tool

logger = logging.getLogger(__name__)


class CodeReviewAgent:
    """
    Automated code review agent

    Provides comprehensive code analysis including:
    - Security vulnerability detection (OWASP compliance)
    - Performance optimization suggestions
    - Code quality and maintainability checks
    - Best practices enforcement

    Example:
        >>> agent = CodeReviewAgent(client)
        >>> report = await agent.review_files(
        ...     workspace="/home/coder/workspace",
        ...     files=["src/api/auth.py", "src/models/user.py"]
        ... )
    """

    def __init__(
        self,
        client: AgentClient,
        model: ModelType = ModelType.CLAUDE_3_5_SONNET,
    ):
        """
        Initialize Code Review Agent

        Args:
            client: AgentClient instance
            model: LLM model to use for reviews
        """
        self.client = client
        self.model = model

    async def review_files(
        self,
        workspace: str,
        files: List[str],
        focus: Optional[List[str]] = None,
    ) -> Dict[str, any]:
        """
        Review code files for issues and improvements

        Args:
            workspace: Absolute workspace path
            files: List of files to review (relative to workspace)
            focus: Optional focus areas (security, performance, quality)

        Returns:
            Dict containing review report with findings
        """
        focus_areas = focus or ["security", "performance", "quality"]
        task = self._build_review_task(files, focus_areas)

        logger.info(f"Starting code review for {len(files)} files")

        request = StartAgentRequest(
            agent_type=AgentType.AIDER,
            workspace=workspace,
            files=files,
            model=self.model,
            task=task,
            metadata={
                "agent_type": "code_review",
                "focus_areas": focus_areas,
                "file_count": len(files),
            },
        )

        agent = await self.client.start_agent(request)
        logger.info(f"Code review agent started: {agent.agent_id}")

        # Collect review findings
        findings = []
        async with self.client.stream_events(agent.agent_id) as stream:
            async for event in stream:
                if event.event.value == "output":
                    findings.append(event.data.get("line", ""))

        return {
            "agent_id": agent.agent_id,
            "status": "completed",
            "files_reviewed": files,
            "focus_areas": focus_areas,
            "findings": findings,
        }

    async def review_pull_request(
        self,
        workspace: str,
        pr_diff: str,
        target_branch: str = "main",
    ) -> Dict[str, any]:
        """
        Review pull request changes

        Args:
            workspace: Absolute workspace path
            pr_diff: Git diff content
            target_branch: Target branch for PR

        Returns:
            Dict containing PR review with inline comments
        """
        task = f"""Review this pull request for merge into {target_branch}.

Analyze:
1. Security vulnerabilities and OWASP compliance
2. Performance implications
3. Code quality and maintainability
4. Breaking changes
5. Test coverage

Provide inline comments with:
- Severity (critical/high/medium/low)
- Description of issue
- Suggested fix

PR Diff:
{pr_diff}
"""

        request = StartAgentRequest(
            agent_type=AgentType.AIDER,
            workspace=workspace,
            model=self.model,
            task=task,
            metadata={
                "agent_type": "pr_review",
                "target_branch": target_branch,
            },
        )

        agent = await self.client.start_agent(request)
        logger.info(f"PR review agent started: {agent.agent_id}")

        comments = []
        async with self.client.stream_events(agent.agent_id) as stream:
            async for event in stream:
                if event.event.value == "output":
                    comments.append(event.data.get("line", ""))

        return {
            "agent_id": agent.agent_id,
            "status": "completed",
            "target_branch": target_branch,
            "comments": comments,
        }

    def _build_review_task(self, files: List[str], focus_areas: List[str]) -> str:
        """Build comprehensive review task description"""
        focus_desc = ", ".join(focus_areas)

        return f"""Perform comprehensive code review focusing on: {focus_desc}

Files to review: {', '.join(files)}

Analysis Requirements:

1. SECURITY ANALYSIS
   - Check for OWASP Top 10 vulnerabilities
   - Identify insecure dependencies
   - Review authentication/authorization logic
   - Check for injection vulnerabilities (SQL, XSS, etc.)
   - Validate input sanitization
   - Review secret management

2. PERFORMANCE ANALYSIS
   - Identify N+1 query patterns
   - Check algorithm complexity (O(n²) or worse)
   - Review memory usage patterns
   - Identify unnecessary computations
   - Check caching opportunities
   - Review database query efficiency

3. CODE QUALITY ANALYSIS
   - Check adherence to SOLID principles
   - Review error handling completeness
   - Validate type safety and null checks
   - Check code duplication (DRY violations)
   - Review function complexity
   - Validate documentation quality

4. BEST PRACTICES
   - Language-specific best practices
   - Framework convention adherence
   - Test coverage adequacy
   - Logging and monitoring

For each issue found, provide:
- File and line number
- Severity (critical/high/medium/low)
- Description of issue
- Code snippet showing the problem
- Recommended fix with code example
- Rationale for the change

Format output as structured JSON for automated processing.
"""


@tool(
    name="analyze_security",
    description="Analyze code for security vulnerabilities",
    tags=["security", "code-review"],
)
async def analyze_security(
    file_path: str,
    workspace: str = "/home/coder/workspace",
) -> Dict[str, any]:
    """
    Analyze file for security vulnerabilities

    Args:
        file_path: Path to file (relative to workspace)
        workspace: Workspace root directory

    Returns:
        Dict containing security analysis results
    """
    full_path = Path(workspace) / file_path

    if not full_path.exists():
        return {"error": f"File not found: {file_path}"}

    # Read file content
    content = full_path.read_text()

    # Security checks (simplified for example)
    findings = []

    # Check for hardcoded secrets
    if any(pattern in content.lower() for pattern in ["password", "api_key", "secret"]):
        findings.append({
            "severity": "high",
            "type": "hardcoded_secret",
            "description": "Potential hardcoded secret detected",
            "line": 0,
        })

    # Check for SQL injection vulnerability
    if "execute(" in content and any(op in content for op in ["+", "f\"{", "%s"]):
        findings.append({
            "severity": "critical",
            "type": "sql_injection",
            "description": "Potential SQL injection vulnerability",
            "line": 0,
        })

    return {
        "file": file_path,
        "findings": findings,
        "scan_date": "2025-10-02",
    }


@tool(
    name="check_performance",
    description="Analyze code for performance issues",
    tags=["performance", "code-review"],
)
async def check_performance(
    file_path: str,
    workspace: str = "/home/coder/workspace",
) -> Dict[str, any]:
    """
    Analyze file for performance issues

    Args:
        file_path: Path to file (relative to workspace)
        workspace: Workspace root directory

    Returns:
        Dict containing performance analysis results
    """
    full_path = Path(workspace) / file_path

    if not full_path.exists():
        return {"error": f"File not found: {file_path}"}

    content = full_path.read_text()
    findings = []

    # Check for nested loops (O(n²))
    if content.count("for ") > 1 or content.count("while ") > 1:
        findings.append({
            "severity": "medium",
            "type": "nested_loop",
            "description": "Nested loops detected - potential O(n²) complexity",
            "line": 0,
        })

    # Check for repeated database calls
    if content.count("query(") > 5:
        findings.append({
            "severity": "high",
            "type": "n_plus_one",
            "description": "Multiple database queries - potential N+1 problem",
            "line": 0,
        })

    return {
        "file": file_path,
        "findings": findings,
        "scan_date": "2025-10-02",
    }

File: src/lib/ai/prompts/templates/code-review.ts
/**
 * Code Review Prompt Templates
 *
 * Comprehensive templates for AI-powered code review covering:
 * - Bug detection
 * - Security vulnerabilities
 * - Performance issues
 * - Code quality and best practices
 * - Style and consistency
 */

import { PromptTemplate, PromptCategory } from '@/types/prompts';

/**
 * Standard code review prompt - comprehensive analysis
 */
export const codeReviewStandard: PromptTemplate = {
  id: 'code-review-standard',
  name: 'Standard Code Review',
  category: PromptCategory.CODE_REVIEW,
  version: '1.0.0',
  description:
    'Comprehensive code review covering bugs, security, performance, and code quality',
  systemPrompt: `You are an expert senior software engineer conducting a thorough code review. Your role is to:

1. Identify potential bugs and logical errors
2. Detect security vulnerabilities and risks
3. Spot performance bottlenecks and inefficiencies
4. Evaluate code quality, readability, and maintainability
5. Check adherence to best practices and coding standards
6. Suggest specific, actionable improvements

Be constructive and educational in your feedback. Explain why something is a problem and how to fix it. Prioritize issues by severity (Critical, High, Medium, Low).`,

  userPromptTemplate: `Please review the following {{language}} code:

\`\`\`{{language}}
{{code}}
\`\`\`

{{#context}}
Additional context:
{{context}}
{{/context}}

{{#focusAreas}}
Please pay special attention to: {{focusAreas}}
{{/focusAreas}}

Provide a structured review with:
1. **Summary**: Brief overview of the code quality
2. **Critical Issues**: Bugs or security vulnerabilities that must be fixed
3. **Performance Concerns**: Any efficiency issues
4. **Code Quality**: Readability, maintainability, and best practices
5. **Suggestions**: Recommended improvements with code examples
6. **Positive Aspects**: What's done well (important for constructive feedback)`,

  variables: [
    {
      name: 'code',
      description: 'The code to review',
      required: true,
      type: 'code',
      example: 'function add(a, b) { return a + b; }'
    },
    {
      name: 'language',
      description: 'Programming language of the code',
      required: true,
      type: 'string',
      defaultValue: 'typescript',
      example: 'typescript'
    },
    {
      name: 'context',
      description: 'Additional context about the code purpose or requirements',
      required: false,
      type: 'string',
      example: 'This is a user authentication function for a banking application'
    },
    {
      name: 'focusAreas',
      description: 'Specific areas to focus on during review',
      required: false,
      type: 'string',
      example: 'security, input validation'
    }
  ],

  recommendedModels: [
    'anthropic/claude-3.5-sonnet',
    'openai/gpt-4o',
    'openai/gpt-4-turbo',
    'google/gemini-pro-1.5'
  ],

  maxTokens: 4096,
  temperature: 0.3,

  modelRequirements: {
    minContextWindow: 8000,
    codeCapable: true,
    capabilities: ['code', 'analysis', 'reasoning']
  },

  modelVariations: [
    {
      modelPattern: 'anthropic/claude-*',
      additionalInstructions:
        'Format your response using clear markdown with headers and code blocks. Use bullet points for lists.'
    },
    {
      modelPattern: 'openai/gpt-4*',
      temperature: 0.2,
      additionalInstructions:
        'Be precise and concise in your analysis. Focus on actionable feedback.'
    }
  ],

  tags: ['review', 'quality', 'security', 'performance', 'bugs'],
  isSystem: true,

  examples: [
    {
      description: 'Review a simple TypeScript function',
      variables: {
        code: `async function fetchUserData(userId: string) {
  const response = await fetch('/api/users/' + userId);
  const data = response.json();
  return data;
}`,
        language: 'typescript',
        context: 'API utility function for fetching user data'
      },
      expectedOutput:
        'Review highlighting missing error handling, potential XSS via userId, missing await on response.json()'
    }
  ]
};

/**
 * Security-focused code review
 */
export const codeReviewSecurity: PromptTemplate = {
  id: 'code-review-security',
  name: 'Security Code Review',
  category: PromptCategory.CODE_REVIEW,
  version: '1.0.0',
  description:
    'Security-focused code review to identify vulnerabilities and security risks',
  systemPrompt: `You are a security expert specializing in application security and secure code review. Your task is to identify security vulnerabilities and risks in code.

Focus on:
- Injection vulnerabilities (SQL, XSS, Command injection, etc.)
- Authentication and authorization flaws
- Sensitive data exposure
- Security misconfigurations
- Cryptographic failures
- Insecure deserialization
- Input validation issues
- Access control problems
- OWASP Top 10 vulnerabilities

Rate each finding by severity: Critical, High, Medium, Low
Provide specific remediation guidance with code examples.`,

  userPromptTemplate: `Perform a security-focused review of this {{language}} code:

\`\`\`{{language}}
{{code}}
\`\`\`

{{#applicationContext}}
Application context: {{applicationContext}}
{{/applicationContext}}

{{#dataHandled}}
Types of data handled: {{dataHandled}}
{{/dataHandled}}

Please provide:
1. **Security Findings**: List all vulnerabilities found, sorted by severity
2. **Risk Assessment**: Potential impact of each vulnerability
3. **Remediation**: Specific fixes with secure code examples
4. **Security Best Practices**: Additional recommendations to improve security posture
5. **Compliance Notes**: Relevant security standards (OWASP, PCI-DSS, etc.)`,

  variables: [
    {
      name: 'code',
      description: 'The code to review for security issues',
      required: true,
      type: 'code'
    },
    {
      name: 'language',
      description: 'Programming language',
      required: true,
      type: 'string',
      defaultValue: 'typescript'
    },
    {
      name: 'applicationContext',
      description: 'Type of application (web, API, mobile, etc.)',
      required: false,
      type: 'string',
      example: 'REST API for e-commerce platform'
    },
    {
      name: 'dataHandled',
      description: 'Types of sensitive data the code handles',
      required: false,
      type: 'string',
      example: 'user credentials, payment information'
    }
  ],

  recommendedModels: [
    'anthropic/claude-3.5-sonnet',
    'openai/gpt-4o',
    'openai/gpt-4-turbo'
  ],

  maxTokens: 4096,
  temperature: 0.2,

  tags: ['security', 'vulnerabilities', 'owasp', 'penetration-testing'],
  isSystem: true
};

/**
 * Performance-focused code review
 */
export const codeReviewPerformance: PromptTemplate = {
  id: 'code-review-performance',
  name: 'Performance Code Review',
  category: PromptCategory.CODE_REVIEW,
  version: '1.0.0',
  description:
    'Performance-focused code review to identify bottlenecks and optimization opportunities',
  systemPrompt: `You are a performance engineering expert. Analyze code for performance issues and optimization opportunities.

Focus on:
- Time complexity and algorithmic efficiency
- Memory usage and potential leaks
- I/O operations and blocking calls
- Database query optimization
- Caching opportunities
- Unnecessary computations or redundant operations
- Async/await patterns and concurrency issues
- Resource management

Provide Big-O analysis where applicable and suggest optimized alternatives.`,

  userPromptTemplate: `Analyze this {{language}} code for performance issues:

\`\`\`{{language}}
{{code}}
\`\`\`

{{#scaleExpectations}}
Expected scale: {{scaleExpectations}}
{{/scaleExpectations}}

{{#performanceCriteria}}
Performance criteria: {{performanceCriteria}}
{{/performanceCriteria}}

Please provide:
1. **Performance Analysis**: Current complexity and bottlenecks
2. **Critical Issues**: Performance problems that need immediate attention
3. **Optimization Opportunities**: Suggested improvements with expected gains
4. **Optimized Code**: Refactored version with performance improvements
5. **Benchmarking Suggestions**: How to measure and verify improvements`,

  variables: [
    {
      name: 'code',
      description: 'The code to analyze for performance',
      required: true,
      type: 'code'
    },
    {
      name: 'language',
      description: 'Programming language',
      required: true,
      type: 'string',
      defaultValue: 'typescript'
    },
    {
      name: 'scaleExpectations',
      description: 'Expected data scale or load',
      required: false,
      type: 'string',
      example: '10,000 requests per second, 1 million records'
    },
    {
      name: 'performanceCriteria',
      description: 'Specific performance requirements',
      required: false,
      type: 'string',
      example: 'response time under 100ms, memory under 512MB'
    }
  ],

  recommendedModels: [
    'anthropic/claude-3.5-sonnet',
    'openai/gpt-4o',
    'google/gemini-pro-1.5'
  ],

  maxTokens: 4096,
  temperature: 0.3,

  tags: ['performance', 'optimization', 'efficiency', 'scalability'],
  isSystem: true
};

/**
 * Quick code review for rapid feedback
 */
export const codeReviewQuick: PromptTemplate = {
  id: 'code-review-quick',
  name: 'Quick Code Review',
  category: PromptCategory.CODE_REVIEW,
  version: '1.0.0',
  description: 'Fast, concise code review for quick feedback on small changes',
  systemPrompt: `You are a code reviewer providing quick, focused feedback. Be concise and prioritize the most important issues. Focus on:
- Obvious bugs
- Clear security issues
- Major code smells
- Simple improvements

Keep your response brief and actionable.`,

  userPromptTemplate: `Quick review of this {{language}} code:

\`\`\`{{language}}
{{code}}
\`\`\`

Provide brief feedback on:
- Any bugs or issues (if found)
- Most important improvement
- Overall assessment (Good/Needs Work/Concerns)`,

  variables: [
    {
      name: 'code',
      description: 'The code to review',
      required: true,
      type: 'code'
    },
    {
      name: 'language',
      description: 'Programming language',
      required: true,
      type: 'string',
      defaultValue: 'typescript'
    }
  ],

  recommendedModels: [
    'openai/gpt-4o-mini',
    'anthropic/claude-3-haiku',
    'google/gemini-flash-1.5'
  ],

  maxTokens: 1024,
  temperature: 0.3,

  tags: ['review', 'quick', 'feedback'],
  isSystem: true
};

/**
 * Pull request review template
 */
export const codeReviewPullRequest: PromptTemplate = {
  id: 'code-review-pull-request',
  name: 'Pull Request Review',
  category: PromptCategory.CODE_REVIEW,
  version: '1.0.0',
  description:
    'Structured review for pull requests with diff context',
  systemPrompt: `You are reviewing a pull request. Analyze the changes in context of:
- The overall change purpose
- How it affects the existing codebase
- Whether it follows project conventions
- Test coverage adequacy
- Documentation needs

Provide inline-style comments where appropriate, referencing specific line changes.`,

  userPromptTemplate: `Review this pull request:

**Title**: {{prTitle}}
**Description**: {{prDescription}}

**Changes**:
\`\`\`diff
{{diffContent}}
\`\`\`

{{#existingCode}}
**Relevant existing code for context**:
\`\`\`{{language}}
{{existingCode}}
\`\`\`
{{/existingCode}}

Please provide:
1. **Summary**: What the PR does and overall assessment
2. **Inline Comments**: Specific feedback on changes (reference line numbers)
3. **Blocking Issues**: Must-fix before merge
4. **Suggestions**: Nice-to-have improvements
5. **Approval Status**: Approve / Request Changes / Need Discussion`,

  variables: [
    {
      name: 'prTitle',
      description: 'Pull request title',
      required: true,
      type: 'string',
      example: 'Add user authentication middleware'
    },
    {
      name: 'prDescription',
      description: 'Pull request description',
      required: true,
      type: 'string'
    },
    {
      name: 'diffContent',
      description: 'The diff/changes in the PR',
      required: true,
      type: 'code'
    },
    {
      name: 'language',
      description: 'Primary programming language',
      required: false,
      type: 'string',
      defaultValue: 'typescript'
    },
    {
      name: 'existingCode',
      description: 'Relevant existing code for context',
      required: false,
      type: 'code'
    }
  ],

  recommendedModels: [
    'anthropic/claude-3.5-sonnet',
    'openai/gpt-4o',
    'openai/gpt-4-turbo'
  ],

  maxTokens: 4096,
  temperature: 0.3,

  tags: ['review', 'pull-request', 'diff', 'git'],
  isSystem: true
};

/**
 * Export all code review templates
 */
export const codeReviewTemplates: PromptTemplate[] = [
  codeReviewStandard,
  codeReviewSecurity,
  codeReviewPerformance,
  codeReviewQuick,
  codeReviewPullRequest
];

export default codeReviewTemplates;
